### DISCUSSION TITLE
WIX 4.0.0 msi does not work on my machine

### DISCUSSION THREAD

**Discussion by NikhatNOV** [wixtoolset/issues#6555] [Questions] [answered]
Used the below code and made an msi using wix build command
The msi does not work on my machine, I says I need a newer version of windows installer. Which version of windows installer do i need

    <?xml version="1.0" encoding="utf-8"?>
    <Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
      <Package Codepage="65001" Manufacturer="Test" ProductCode="{6988DDD6-9DB6-48D4-A41A-B81346380831}" Language="1033" Name="Test MSI" Version="1.0.0" UpgradeCode="{CA37CA48-F0B4-4BB6-8686-929601E0C7E6}" InstallerVersion="1" Compressed="yes">
        <StandardDirectory Id="ProgramFilesFolder">
          <Directory Id="INSTALLFOLDER" Name="NOV">
            <Directory Id="HostAppDir" ShortName="fcnnflwf" Name="Test">
              <Component Id="cmp883DA1B24EE5E0CCCBCDABFE6C563865" Guid="{469BDD74-043A-4061-B745-63B2D57EA526}" Bitness="always32">
                <File Id="filFB30554DBD98546A19E0BEDC6A6A6753"  Name="natsconf" KeyPath="yes" Source="C:\TEST\nats-streaming-server-v0.22.1-windows-amd64\nats-streaming-server.conf" />
              </Component>
            </Directory>
          </Directory>
        </StandardDirectory>
        <Feature Id="ProductFeature" Title="Test" Level="1">
          <ComponentRef Id="cmp883DA1B24EE5E0CCCBCDABFE6C563865" />
        </Feature>
        <Media Id="1" Cabinet="cab1.cab" />
        <SummaryInformation Description="NOV Test  v1.0" />
        <MajorUpgrade DowngradeErrorMessage="A newer version." />
      </Package>
    </Wix>

**Comment by barnson** [comment]
Please answer the questions in the bug template, including the version of Windows you're running on.

**Comment by jhennessey** [comment, marked as answer]
@NikhatNOV - It appears that you specified an invalid `InstallerVersion` value. See the documentation for the attribute here: https://wixtoolset.org/documentation/manual/v4/reference/wxs/package/

Assuming it actually built with your provided code, you could argue an error should have been produced.

**Comment by NikhatNOV** [comment]
@jhennessey : Thank you so much. Now the installer builds and installs the files. I find the documentation quite difficult to follow. Do you have any examples to install exe's as service. And also uninstall them when msi is uninstalled.

I have tried to write this but it fails often. Not always.
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <Package Codepage="65001" Manufacturer="Test" ProductCode="{6988DDD6-9DB6-48D4-A41A-B81346380831}" Language="1033" Name="Test" Version="1.0.0" UpgradeCode="{CA37CA48-F0B4-4BB6-8686-929601E0C7E6}" InstallerVersion="200" Compressed="yes">
    <StandardDirectory Id="ProgramFilesFolder">
      <Directory Id="INSTALLFOLDER" Name="test">
        <Directory Id="HostAppDir" ShortName="fcnnflwf" Name="Test">
          <Component Id="cmp469BDD74043A4061B74563B2D57EA526" Guid="{469BDD74-043A-4061-B745-63B2D57EA526}" Bitness="always32">
            <File Id="fil469BDD74043A4061B74563B2D57EA526"  Name="nats-streaming-server.conf" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\nats-streaming-server.conf" />
          </Component>
          <Component Id="cmpC95C5FD3CB9C4A898F4E0D8E32741025" Guid="{C95C5FD3-CB9C-4A89-8F4E-0D8E32741025}" Bitness="always32">
            <File Id="filC95C5FD3CB9C4A898F4E0D8E32741025"  Name="nats-server.conf" KeyPath="yes" Source="C:\Test\nats-streaming-server-v0.22.1-windows-amd64\nats-server.conf" />
          </Component>
          <Component Id="cmpF2E694600BFA4B01803781552006078F" Guid="{F2E69460-0BFA-4B01-8037-81552006078F}" Bitness="always32">
            <File Id="filF2E694600BFA4B01803781552006078F"  Name="maxedge-core-messagebus.exe" KeyPath="yes" Source="C:\Test\nats-streaming-server-v0.22.1-windows-amd64\maxedge-core-messagebus.exe" />
          </Component>
          <Component Id="cmpF87A99D4F16247FC8EF48FDB0361968B" Guid="{F87A99D4-F162-47FC-8EF4-8FDB0361968B}" Bitness="always32">
            <File Id="filF87A99D4F16247FC8EF48FDB0361968B"  Name="InstallNATS.cmd" KeyPath="yes" Source="C:\Test\nats-streaming-server-v0.22.1-windows-amd64\InstallNATS.cmd" />
          </Component>
        </Directory>
      </Directory>
    </StandardDirectory>
    <Feature Id="ProductFeature" Title="Test" Level="1">
      <ComponentRef Id="cmp469BDD74043A4061B74563B2D57EA526" />
      <ComponentRef Id="cmpC95C5FD3CB9C4A898F4E0D8E32741025" />
      <ComponentRef Id="cmpF2E694600BFA4B01803781552006078F" />
      <ComponentRef Id="cmpF87A99D4F16247FC8EF48FDB0361968B" />
    </Feature>
    <CustomAction Id="RunNATSServiceInstall" Directory="HostAppDir" ExeCommand="[HostAppDir]InstallNATS.cmd" Execute="deferred" Return="check" />
    <CustomAction Id="RunNATSServiceStop" Directory="HostAppDir" ExeCommand="[HostAppDir]maxedge-core-messagebus.exe stop" Execute="deferred" Return="check" />
    <CustomAction Id="RunNATSServiceUninstall" Directory="HostAppDir" ExeCommand="[HostAppDir]maxedge-core-messagebus.exe uninstall" Execute="deferred" Return="check" />
     <InstallExecuteSequence>
        
        <Custom Action="RunNATSServiceInstall" After="WriteIniValues"></Custom>
      </InstallExecuteSequence>
    <SummaryInformation Description="Test  v1.0" />
    <MajorUpgrade DowngradeErrorMessage="A newer version of NTest is already installed." />
  </Package>
</Wix>

**Comment by BMurri** [comment]
> I find the documentation quite difficult to follow. Do you have any examples to install exe's as service. And also uninstall them when msi is uninstalled.

The `How-Tos` and other non-syntax documentation have not yet been migrated from v3, so you will have to adapt as appropriate, but the concepts still (mostly) apply.

In your case, look [here](https://wixtoolset.org/documentation/manual/v3/howtos/general/install_windows_service.html).

**Comment by NikhatNOV** [comment]
I am still unable to uninstall a service, This custom action somehow does not get executed.<Custom Action="RunNATSServiceUninstall" After="InstallInitialize"></Custom>

This is my code

```
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <Package Codepage="65001" Manufacturer="TEST" ProductCode="{6988DDD6-9DB6-48D4-A41A-B81346380831}" Language="1033" Name="Test" Version="1.0.0" UpgradeCode="{CA37CA48-F0B4-4BB6-8686-929601E0C7E6}" InstallerVersion="200" Compressed="yes">
      <Directory Id="TARGETDIR" Name="SourceDir">
      <Directory Id="TEST" Name="TEST">
        <Directory Id="HostAppDir" ShortName="fcnnflwf" Name="Test">
          <Component Id="cmp469BDD74043A4061B74563B2D57EA526" Guid="{469BDD74-043A-4061-B745-63B2D57EA526}" Bitness="always32">
            <File Id="fil469BDD74043A4061B74563B2D57EA526"  Name="nats-streaming-server.conf" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\nats-streaming-server.conf" />
          </Component>
          <Component Id="cmpC95C5FD3CB9C4A898F4E0D8E32741025" Guid="{C95C5FD3-CB9C-4A89-8F4E-0D8E32741025}" Bitness="always32">
            <File Id="filC95C5FD3CB9C4A898F4E0D8E32741025"  Name="nats-server.conf" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\nats-server.conf" />
          </Component>
          <Component Id="cmpF2E694600BFA4B01803781552006078F" Guid="{F2E69460-0BFA-4B01-8037-81552006078F}" Bitness="always32">
            <File Id="filF2E694600BFA4B01803781552006078F"  Name="core-messagebus.exe" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\core-messagebus.exe" />
          </Component>
          <Component Id="cmpF87A99D4F16247FC8EF48FDB0361968B" Guid="{F87A99D4-F162-47FC-8EF4-8FDB0361968B}" Bitness="always32">
            <File Id="filF87A99D4F16247FC8EF48FDB0361968B"  Name="InstallNATS.cmd" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\InstallNATS.cmd" />
          </Component>
  	<Component Id="cmp3AD1681AB207412AB7E6862B334A423A" Guid="{3AD1681A-B207-412A-B7E6-862B334A423A}" Bitness="always32">
            <File Id="fil3AD1681AB207412AB7E6862B334A423A"  Name="UninstallNATS.cmd" KeyPath="yes" Source="C:\OnWindows\nats-streaming-server-v0.22.1-windows-amd64\UninstallNATS.cmd" />
          </Component>
        </Directory>
      </Directory>
    </Directory>
    <Feature Id="ProductFeature" Title="Test" Level="1">
      <ComponentRef Id="cmp469BDD74043A4061B74563B2D57EA526" />
      <ComponentRef Id="cmpC95C5FD3CB9C4A898F4E0D8E32741025" />
      <ComponentRef Id="cmpF2E694600BFA4B01803781552006078F" />
      <ComponentRef Id="cmpF87A99D4F16247FC8EF48FDB0361968B" />
      <ComponentRef Id="cmp3AD1681AB207412AB7E6862B334A423A"  />
    </Feature>
    <CustomAction Id="RunNATSServiceInstall" Directory="HostAppDir" ExeCommand="[HostAppDir]InstallNATS.cmd" Execute="deferred" Return="check" />
    <CustomAction Id="RunNATSServiceUninstall" Directory="HostAppDir" ExeCommand="[HostAppDir]UninstallNATS.cmd" Execute="deferred" Return="check" />
     <!--Changes the default installation location to C:\-->
     <CustomAction Id="SetRootDrive" Property="ROOTDRIVE" Value="[%SystemDrive]\" />
     <InstallUISequence>
      <Custom Action="SetRootDrive" Before="CostInitialize" />
     </InstallUISequence>
     <InstallExecuteSequence>
       <Custom Action="SetRootDrive" Before="CostInitialize" />
     </InstallExecuteSequence>     
     <InstallExecuteSequence>
        <!--ApiServices InstallExecute-->
        <Custom Action="RunNATSServiceInstall" After="WriteIniValues"></Custom>
	<Custom Action="RunNATSServiceUninstall" After="InstallInitialize"></Custom>
      </InstallExecuteSequence>
    <SummaryInformation Description="" />
    <MajorUpgrade DowngradeErrorMessage="A newer version of Test is already installed." />
  </Package>
</Wix>
```

**Comment by barnson** [comment]
We moved this issue to the [experimental GitHub discussions area][gh]. You might also consider asking your question on the [wix-users mailing list][ml] or posting on [Stack Overflow][so].

[gh]: https://github.com/wixtoolset/issues/discussions?discussions_q=category%3AQuestions
[ml]: http://wixtoolset.org/documentation/mailinglist/
[so]: https://stackoverflow.com/questions/tagged/wix
